tests/test-class-wc-payment-gateway-wcpay-subscriptions-process-payment.php

```diff
@@ -0,0 +1,359 @@
+			->willReturn( $this->payment_intent );
+
+		$this->mock_token_service
+			->expects( $this->never() )
+			->method( 'add_payment_method_to_user' );
+
+		// Expect add token to order to be called, so it can be reused in renewals.
+		$this->mock_wcpay_gateway
+			->expects( $this->once() )
+			->method( 'add_token_to_order' )
+			->with(
+				$this->callback( $this->match_order_id( $order->get_id() ) ),
+				$this->token
+			);
+
+		$result       = $this->mock_wcpay_gateway->process_payment( $order->get_id() );
+		$result_order = wc_get_order( $order->get_id() );
+
+		$this->assertEquals( 'processing', $result_order->get_status() );
+		$this->assertEquals( 'success', $result['result'] );
+	}
+
+	public function test_saved_card_zero_dollar_subscription() {
+		$order = WC_Helper_Order::create_order( self::USER_ID, 0 );
+
+		$_POST = [ self::TOKEN_REQUEST_KEY => $this->token->get_id() ];
+
+		$this->mock_wcs_order_contains_subscription( true );
+
+		$this->mock_api_client
+			->expects( $this->once() )
+			->method( 'create_setup_intent' )
+			->with( self::PAYMENT_METHOD_ID, self::CUSTOMER_ID, 'true' )
+			->willReturn( $this->setup_intent );
+
+		$this->mock_token_service
+			->expects( $this->never() )
+			->method( 'add_payment_method_to_user' );
+
+		// Expect add token to order to be called, so it can be reused in renewals.
+		$this->mock_wcpay_gateway
+			->expects( $this->once() )
+			->method( 'add_token_to_order' )
+			->with(
+				$this->callback( $this->match_order_id( $order->get_id() ) ),
+				$this->token
+			);
+
+		$result       = $this->mock_wcpay_gateway->process_payment( $order->get_id() );
+		$result_order = wc_get_order( $order->get_id() );
+
+		$this->assertEquals( 'processing', $result_order->get_status() );
+		$this->assertEquals( 'success', $result['result'] );
+	}
+
+	public function test_saved_card_is_added_before_status_update() {
+		$order = WC_Helper_Order::create_order( self::USER_ID, 0 );
+
+		$_POST = [ self::TOKEN_REQUEST_KEY => $this->token->get_id() ];
+
+		$this->mock_wcs_order_contains_subscription( true );
+
+		$this->mock_api_client
+			->expects( $this->once() )
+			->method( 'create_setup_intent' )
+			->with( self::PAYMENT_METHOD_ID, self::CUSTOMER_ID, 'true' )
+			->willReturn( $this->setup_intent );
+
+		$this->mock_token_service
+			->expects( $this->never() )
+			->method( 'add_payment_method_to_user' );
+
+		// Expect add token to order to be called, so it can be reused in renewals.
+		$this->mock_wcpay_gateway
+			->expects( $this->once() )
+			->method( 'add_token_to_order' )
+			->with(
+				$this->logicalAnd(
+					$this->callback( $this->match_order_id( $order->get_id() ) ),
+					$this->callback( $this->match_order_status( 'pending' ) )
+				),
+				$this->token
+			);
+
+		$result = $this->mock_wcpay_gateway->process_payment( $order->get_id() );
+	}
```

Comment:
I feel like we're missing assertions here again. Or am I just missing something?

Reply:
This is similar to the comment above, we're asserting when and how `add_token_to_order` was called. Do you think we should include assertions on the final order status?

Reply:
Nah, I just missed the `expects` call 🙂. I think it's good as it is!